### ISSUE TITLE
Add a j2objc-maven-plugin to integrate Objective-C code generation into Maven build lifecycle [moved]

### ISSUE THREAD

**Issue by kstanger** [google/j2objc#160]
This is [Issue 160](https://code.google.com/p/j2objc/issues/detail?id=160) moved from a Google Code project.
Added on 2013-04-01T13:17:45.000Z by [cristian...@gmail.com](http://code.google.comhttps://code.google.com/u/111428147050763296727/).
Please review that bug for more context and additional comments, but update this bug.

Original labels: Type-Enhancement, Priority-Medium

**Comment by ArloL** [comment]
I tried using the [`executable`](http://maven.apache.org/plugins/maven-compiler-plugin/compile-mojo.html#executable) and [`fork`](http://maven.apache.org/plugins/maven-compiler-plugin/compile-mojo.html#executable) configuration of the [maven-compiler-plugin](http://maven.apache.org/plugins/maven-compiler-plugin/compile-mojo.html) but sadly it did not work. See https://github.com/ArloL/maven-quickstart-j2objc for a quickstart project that I created demonstrating the problem. Correct the path in the settings.xml and call `run.sh`.

The important part of the console output:

~~~~
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.2:compile (default-compile) on project test: Compilation failure: Compilation failure:
[ERROR] error: Unknown file type: "-d"
[ERROR] error: Unknown file type: "/Users/ArloLOK/Developer/j2objc/test/target/classes"
[ERROR] error: Unknown file type: "-classpath"
[ERROR] error: Unknown file type: "/Users/ArloLOK/Developer/j2objc/test/target/classes:"
[ERROR] error: Unknown file type: "-sourcepath"
[ERROR] error: Unknown file type: "/Users/ArloLOK/Developer/j2objc/test/src/main/java:/Users/ArloLOK/Developer/j2objc/test/target/generated-sources/annotations:"
[ERROR] error: Unknown file type: "/Users/ArloLOK/Developer/j2objc/test/src/main/java/de/test/App.java"
[ERROR] error: Unknown file type: "-g"
[ERROR] error: Unknown file type: "-nowarn"
[ERROR] error: Unknown file type: "-target"
[ERROR] error: Unknown file type: "1.5"
[ERROR] error: Unknown file type: "-source"
[ERROR] error: Unknown file type: "1.5"
[ERROR] error: Unknown file type: "-encoding"
[ERROR] error: Unknown file type: "UTF-8"
~~~~

**Comment by tomball** [comment]
The maven-compiler-plugin looks like it would be useful, but according to [its development guide](http://maven.apache.org/plugins/maven-compiler-plugin/non-javac-compilers.html), only compilers its supports will work. Since j2objc's arguments are similar to javac's, adding j2objc to maven-compiler-plugin's external compiler list may work.

**Comment by tomball** [comment]
It looks like [plexus-compilers](https://github.com/codehaus-plexus/plexus-compiler) is the right project to hack.

**Comment by rastaman** [comment]
Hi, is there any progress on this issue ? I'm not afraid to look at it, but would like to know if someone have already worked on it (the hint from tomball looks good).
Best regards

**Comment by tomball** [comment]
As far as we know, no one has started working on this. The Google engineers
are in the middle of a big build system update (moving to Bazel
<http://bazel.io/>*), and so don't have the bandwidth to improve support
for other build systems currently. Please let us know if you have any
success!

Tom

*We don't recommend anyone move to Bazel yet, as there needs to be custom
j2objc support added before it's useful for other projects.

**Comment by rastaman** [comment]
Hi, i've done a proof of concept ( https://github.com/rastaman/plexus-compiler/tree/master/plexus-compilers/plexus-compiler-j2objc ) and will submit it to plexus-compiler ( https://github.com/codehaus-plexus/plexus-compiler/issues/1 ) when it will be more usable.

**Comment by tomball** [comment]
That's exciting news. Let us know if plexus-compiler accepts it, so we can
add a "building with maven" wiki page (or contribute one you've written, if
you'd prefer).

**Comment by rastaman** [comment]
Hi! I've submitted the pull request to plexus-compiler. Meanwhile, anybody can try the compiler and report any bugs to me. I've forked the maven quickstart project here (with a bigger model) so any one can try and see how to use the compiler :  https://github.com/rastaman/maven-quickstart-j2objc

**Comment by rastaman** [comment]
Hi! The pull request have been accepted by plexus-compiler ( https://github.com/codehaus-plexus/plexus-compiler.git / https://github.com/codehaus-plexus/plexus-compiler/tree/master/plexus-compilers/plexus-compiler-j2objc ). There is a minimal documentation in the README.md file.
Have a good day.

**Comment by rastaman** [comment]
The pull request to add a J2ObjC compiler has been merged in plexus-compiler, if anyone want to elaborate on the documentation, feel free to do it ! :-)
Best regards

**Comment by ryanproud** [comment]
@rastaman Not sure if I'm missing some config, but 2.6-SNAPSHOT doesn't appear to be available and the quickstart project fails with 2.6.

> Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.3:compile (default-compile) on project test: Fatal error compiling: basedir maven-quickstart-j2objc/target/generated-sources/annotations does not exist

**Comment by rastaman** [comment]
Hi @ryanproud , 
I have tried the maven-quickstart-j2objc project with the current github master branch of plexus-compiler (2.7-SNAPSHOT) and it was failing like you describe. I have added the following conf to the maven  compiler plugin configuration :
```
	<generatedSourcesDirectory>src/main/java</generatedSourcesDirectory>
```
And it works. I have pushed the change in the repo of the quickstart project. If it didn't works after this patch, pleace open an issue inside the project and discuss it from here.

Thanks for your reporting, best regards.
